### DISCUSSION TITLE
Dynu Bad authentication

### DISCUSSION THREAD

**Discussion by flixxx** [qdm12/ddns-updater#385] [General]
Trying to setup ddns-updater with my provider dynu
logs pulls the right info (current ip, ip logged on their site, domain, etc..)
when it comes time to update it, the following error;
ERROR bad authentication

I tried using my password, ip updater password, md5 hashed password.

**Comment by flixxx** [comment]
I've been messing around, tested dynu's api and my information works properly. I believe ddns-updater may not be sending the right formatting.
how can we check what it sends to dynu? Is there a verbose log available.

**Reply by qdm12** [reply]
Try `LOG_LEVEL=debug` it should log all requests and responses

**Reply by qdm12** [reply]
Hello, did you have time to debug it? Thanks!

**Reply by JackTalisker** [reply]
I have the same problem.
Tried both user and email address as username.
The debug log shows the following:
```

2022/12/03 14:11:56 INFO Updating record [domain: MYDOMAIN.com | host: MYHOST | provider: dynu | ip: ipv4] to use 123.191.135.321
2022/12/03 14:11:56 DEBUG GET https://api.dynu.com/nic/update?alias=MYALIAS&hostname=MYHOSTNAME.com&location=&password=MYPASSWORD&username=MYUSERNAME%40gmail.com | headers: User-Agent: DDNS-Updater quentin.mcgaw@gmail.com
2022/12/03 14:11:59 DEBUG 200 OK | headers: Date: Sat, 03 Dec 2022 13:11:59 GMT; Server: Dynu Web Server; X-Powered-By: Dynu Dynamic DNS Service; Content-Length: 7; Content-Type: text/html; charset=UTF-8 | body: badauth
2022/12/03 14:11:59 ERROR bad authentication

```

**Reply by JackTalisker** [reply]
I solved this way:
my old config was
```
    {
      "provider": "dynu",
      "domain": "mydomain.com",
      "host": "myhost",
      "username": "myusername",
      "password": "mypassword",
      "ip_version": "ipv4",
      "provider_ip": true
    }

```
I changed to:
```
    {
      "provider": "dynu",
      "domain": "myhost.mydomain.com",
      "host": "@",
      "username": "myusername",
      "password": "mypassword",
      "ip_version": "ipv4",
      "provider_ip": true
    }

```

now it works but I suppose it's only a workaround (maybe the problem is the "alias" handling?)

**Reply by qdm12** [reply]
Awesome that helped quite a bit. Also sorry for the mega delay.

**Comment by qdm12** [comment]
Fixed in 83a9a2d999392dfcaee1d0b0e3dab538d52c337c

**Reply by vespian** [reply]
@qdm12 `domain`   field has different semantics depending how you use it in the API. https://github.com/qdm12/ddns-updater/commit/83a9a2d999392dfcaee1d0b0e3dab538d52c337c  in its current form results API in badauth response/broke Dynu provider:

```
curl https://api.dynu.com/nic/update\?hostname\=kato.ext.vespian.net\&myip\=2.23.3.2\&password\=...\&username\=...
badauth
```

Adding a dot to the hostname makes it work (which BTW is forbidden by validation checks in the config), and the root of the domain is updated, not the subhost which is probably what @JackTalisker was after:

```
curl https://api.dynu.com/nic/update\?hostname\=kato.ext.vespian.net.\&myip\=2.23.3.2\&password\=...\&username\=... 
nochg
```

When using alias + hostname, it works and it updates exactly one entry:

```
curl https://api.dynu.com/nic/update\?alias\=kato.ext\&hostname\=vespian.net\&myip\=2.23.3.2\&password\=...\&username\=...
nochg
```

Can we please revert https://github.com/qdm12/ddns-updater/commit/83a9a2d999392dfcaee1d0b0e3dab538d52c337c and document the workaround that @JackTalisker was using? 

I have built a fork to confirm it is working, can submit it tomorrow-ish too.

**Reply by qdm12** [reply]
Correct me if I'm wrong:

- update `kato.ext.vespian.net`: set hostname to `vespian.net` + alias to `kato.ext`
- update `ext.vespian.net`: set hostname to `ext.vespian.net`, do not set alias
- update `vespian.net`: set hostname to `vespian.net`, do not set alias


I would prefer to code these instead of documenting this strange API behavior ideally 😉

**Reply by cross** [reply]
I am getting here while searching for similar problems with another ddns updater with dynu.com.  If I submit `hostname=host.domain.com`, given that the "DDNS Domain" I have listed with dynu.com is "domain.com", it will respond "badauth".
If I send `hostname=domain.com&alias=host` it will successfully set the host A record inside of domain.com to the specified address.  As noted by @vespian , this would have been the behavior before the change suggested by @JackTalisker .
Jack, what is the domain listed in the Dynu control panel table for you?  Is it "mydomain.com", or "myhost.mydomain.com" ?

**Reply by qdm12** [reply]
@JackTalisker (gentle reminder 😉)

**Reply by JackTalisker** [reply]
Here is my control panel. How can I help?
![Screenshot 2024-09-07 210330](https://github.com/user-attachments/assets/266fa038-1697-4b50-9944-1c98305d6ff0)

**Reply by cross** [reply]
I was describing the previous page, but I think but I think the link there opens the above page, and the information is what you masked in the upper left.
I don't need to know the exact string there, but I wanted to know if that is the exact string you are setting an address for.  In my case, that string there (the domain) is "distal.com", and the names I want to set are _inside_ of it, in the DNS Records link on the right hand side of the DDNS control panel.
Does the question make sense?  Is the address you're setting with ddns-updater the same as the string there at the top, or a FQDN a level under that?

**Comment by JackTalisker** [comment]
Here is my DNS Records page

![Screenshot 2024-09-08 083348](https://github.com/user-attachments/assets/c40ea71f-8eba-4439-b4c6-b71a5c215945)

**Reply by cross** [reply]
Thanks, Jack.  That confirms my suspicion, that you are setting the address of the primary "domain" you have set up with dynu.  You have only the root and wildcard records beneath it, both set to the same value (I presume).
So, in that case, `hostname` is to be used, and apparently it needs to be without an `alias` ?  Responding to @vespian comment, Jack was in fact after the root of the domain, not a subhost.
So we need the hostname/alias syntax when the root domain dynu knows is not the same as the address being set.  When you are updating the domain itself, you can update the hostname itself.  (and Jack has shown that an "@" as alias works, but is not needed and possibly confusing?)
IMO, @JackTalisker 's original reconfiguration "workaround" is actually correct.  It's just confusing given the keys in config.  He doesn't want to set the hostname, he wants to set the root domain.  That "domain" is just also being used as a hostname.  
In response to @qdm12 comment, "it depends on what you have registered at dynu".  I am comfortable with "domain" needing to be what's registered, and if you need a member of that domain set, rather than the domain itself, you specify that as a "hostname" parameter.  It just is confusing if you have registered a "domain" that in your own opinion is a hostname.
